extensions/amp-analytics/0.1/events.js

```diff
@@ -389,6 +395,43 @@ export class VisibilityTracker extends EventTracker {
   }
 
   /**
+   * @param {string=} waitForSpec
+   * @param {string=} selector
+   * @param {Element=} element
+   * @return {?Promise}
+   * @visibleForTesting
+   */
+  getReadyPromise(waitForSpec, selector, element) {
+    if (waitForSpec == WAITFOR_EVENTS_LIST.NONE || !selector) {
+      // Always return null if specific asked waitFor NONE, or no selector
+      return null;
+    }
+    if (!element) {
+      // If selector is :root or :host, always wait for root ini-load
+      return this.iniLoadTracker_.getRootSignal();
+    }
+    if (!waitForSpec) {
+      // Default behavior for selector select AMP element is to wait for
+      // AMP element's ini-load
+      waitForSpec = WAITFOR_EVENTS_LIST.INI_LOAD;
+    }
+    if (typeof element.signals != 'function') {
+      return null;
+    }
+    switch (waitForSpec) {
+      case WAITFOR_EVENTS_LIST.INI_LOAD:
+        return this.iniLoadTracker_.getElementSignal(element);
+        break;
+      case WAITFOR_EVENTS_LIST.RENDER_START:
+        // may never resolve.
+        return element.signals().whenSignal(WAITFOR_EVENTS_LIST.RENDER_START);
```

Comment:
We shouldn't proceed (and I don't think we do) when `layoutCallback` fails. However, we have a nuance when we proceed on `no-content`. That should be fixed via another issue.